src/function/scalar/variant/variant_extract.cpp

```diff
@@ -0,0 +1,214 @@
+		                      path.return_type.ToString());
+	}
+
+	Value constant_arg;
+	if (!GetConstantArgument(context, path, constant_arg)) {
+		throw BinderException("'variant_extract' expects the second argument to be a constant expression");
+	}
+
+	if (constant_arg.type().id() == LogicalTypeId::VARCHAR) {
+		return make_uniq<BindData>(constant_arg.GetValue<string>());
+	} else if (constant_arg.type().id() == LogicalTypeId::UINTEGER) {
+		return make_uniq<BindData>(constant_arg.GetValue<uint32_t>());
+	} else {
+		throw InternalException("Constant-folded argument was not of type UINTEGER or VARCHAR");
+	}
+}
+
+//! FIXME: it could make sense to allow a third argument: 'default'
+//! This can currently be achieved with COALESCE(TRY(<extract method>), 'default')
+static void VariantExtractFunction(DataChunk &input, ExpressionState &state, Vector &result) {
+	auto count = input.size();
+
+	D_ASSERT(input.ColumnCount() == 2);
+	auto &variant = input.data[0];
+	D_ASSERT(variant.GetType() == LogicalType::VARIANT());
+
+	auto &path = input.data[1];
+	D_ASSERT(path.GetVectorType() == VectorType::CONSTANT_VECTOR);
+
+	auto &func_expr = state.expr.Cast<BoundFunctionExpression>();
+	auto &info = func_expr.bind_info->Cast<BindData>();
+	auto &allocator = Allocator::DefaultAllocator();
+
+	RecursiveUnifiedVectorFormat source_format;
+	Vector::RecursiveToUnifiedFormat(variant, count, source_format);
+
+	//! Path either contains array indices or object keys
+	SelectionVector value_index_sel;
+	value_index_sel.Initialize(count);
+	for (idx_t i = 0; i < count; i++) {
+		value_index_sel[i] = 0;
+	}
+
+	SelectionVector new_value_index_sel;
+	new_value_index_sel.Initialize(count);
+
+	auto owned_nested_data = allocator.Allocate(sizeof(VariantNestedData) * count);
+	auto nested_data = reinterpret_cast<VariantNestedData *>(owned_nested_data.get());
+
+	string error;
+	auto &component = info.component;
+
+	auto expected_type = component.lookup_mode == VariantChildLookupMode::BY_INDEX ? VariantLogicalType::ARRAY
+	                                                                               : VariantLogicalType::OBJECT;
+	if (!VariantUtils::CollectNestedData(source_format, expected_type, value_index_sel, count, optional_idx(), 0,
+	                                     nested_data, FlatVector::Validity(result), error)) {
+		throw InvalidInputException(error);
+	}
+
+	if (!VariantUtils::FindChildValues(source_format, component, optional_idx(), new_value_index_sel, nested_data,
+	                                   count)) {
+		switch (component.lookup_mode) {
+		case VariantChildLookupMode::BY_INDEX: {
+			throw InvalidInputException("ARRAY does not contain the index: %d", component.index);
+		}
+		case VariantChildLookupMode::BY_KEY: {
+			throw InvalidInputException("OBJECT does not contain the key: %s", component.key);
+		}
+		}
+	}
+
+	//! We have these indices left, the simplest way we can finalize this is:
```

Comment:
Improve this comment, mention that we use the `new_value_index_sel` to create a dictionary on the `values` list entry.
Everything else is untouched, which makes this method essentially zero-copy (if we ignore nulls, because those make this less elegant)